---
title: "R Notebook"
output: github_document
---

```{r}
library(dada2)
```

```{r}
path <- "~/TUTORIEL-ADM/MiSeq_SOP"
list.files(path)
```

  Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq

```{r}
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
```

  # Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
  
```{r}
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

Inspect read quality profiles

```{r, cache=TRUE}
plotQualityProfile(fnFs[1:2])
```

```{r, cache=TRUE}
plotQualityProfile(fnRs[1:2])
```

Filtrer et Rogner

```{r, cache=TRUE}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

Paramètre de base

```{r, cache=TRUE}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=FALSE)

head(out)
```

Apprenez les taux d'erreur

```{r, cache=TRUE}
errF <- learnErrors(filtFs, multithread=TRUE)
```
```{r, cache=TRUE}
errR <- learnErrors(filtRs, multithread=TRUE)
```
Visualisation des taux d'erreur

```{r, cache=TRUE}
plotErrors(errF, nominalQ=TRUE)
```
Application de l'algo de DADA2 

```{r, cache=TRUE}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```

```{r, cache=TRUE}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```

Inspection 

```{r}
dadaFs[[1]]
```
Fusion des lectures appariées

```{r, cache=TRUE}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
```

Inspect the merger from the first list

```{r, cache=TRUE}
head(mergers[[1]])
```

Construction d'une table de séquence

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

  Inspect distribution of sequence lenghs
  
```{r}
table(nchar(getSequences(seqtab)))
```

Supprimer les chimères

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```
```{r}
sum(seqtab.nochim)/sum(seqtab)
```

Track reads through the pipeline

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

Assign taxonomy

```{r, cache=TRUE}
taxa <- assignTaxonomy(seqtab.nochim, "~/TUTORIEL-ADM/silva_nr99_v138.2_toGenus_trainset.fa.gz?download=1", multithread=TRUE)
```

Examination 

```{r, cache=TRUE}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

Evaluation de la précision

```{r}
unqs.mock <- seqtab.nochim["Mock",]
unqs.mock <- sort(unqs.mock[unqs.mock>0], decreasing=TRUE) # Drop ASVs absent in the Mock
cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")
```

```{r}
mock.ref <- getSequences(file.path(path, "HMP_MOCK.v35.fasta"))
match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

